Agent+config support together with Marta's imageVector support and other fixes (unitests)

ppalucki opened the pull request:
Supersedes #27 

More description in above PR and following issues:

Resolves #1, resolves #2, resolves #3, resolves #11 and possibly resolves #28

Changes/features:

configs:
- deploy CRI-RM agent and it Adjustments CRDs - more docs [here](https://intel.github.io/cri-resource-manager/stable/docs/setup.html#container-adjustments)
- dynamic configuration update by agent - docs [here](https://intel.github.io/cri-resource-manager/stable/docs/node-agent.html) 
- configurations files (default/force/per node/per group) can be passed through ControllerDeployment as well as by providerConfig #11  (handle force.cfg - when provided will overwrite fallback and agent )


fixes:
- fix issues with imageVector 

other enhacmentes:
- update gardener to 1.56.0 and cri-rm to 0.7.2
- remove "remove daemonset" and replace it with trap for SIGTERM
- rename targets in Makefile: build-images and push-images optimization (caching) 
- Makefile: replace VERSION with TAG
- replace v2.isvimgreg.com with kind-local-provider localhost:5001
- refactor code (restructure) - move actuator/configs/healthchecks to separate packages,
- add some tests for configs/actuator 

TODO (after Marta review)
- [x] document default values from charts.yaml as examples to tested outside extension (e.g. helm template) no to be commited to repo
- [x] move charts to internal
- [x] consider healthcheck to pkg/controller/ and actuator to pkg/controller/lifecycle 
- [x] remove TODOs/comments 

- [x] enable logging for node-agent (need addtional annotation/label from Damin PR)
- [x] add license headers
- [x] disable zaplogger development option to False

Not done, moved out of this PR:
- consider removing fallback.cfg.sample as default - because it enambles TopologyAware policy which is different from built-in
- tests on SLES
- test what happens when you provide broken config

mmucek95 commented on examples/ctrldeploy-ctrlreg.yaml:
```diff
@@ -5,22 +5,52 @@ metadata:
-  #     repository: v2.isvimgreg.com/gardener-extension-cri-resmgr
-  #     tag: latest
-  #     pullPolicy: Always
+  chart: H4sIAAAAAAAAA+0ca2/bODKf/SsI5xZoi1jx2wfdl3MTb+u75gEnbdFPBS3Rti6SqCUpu952//vNkJKsOI5fSd3urgZFLInkzHBenKGojqlwWchEhX1RLJQeDyuO8CqCyWAsTo+eBaoAnVZL/wIs/+rrWrNRq7fbzVa1eVSttZqN+hFpPQ/59RBLRQUhR4Jzta7fpvY/KYzX6v9sQoWy5jTwn0IDFdxuNh/TP6oe9F9vdWrVRqdWB/13Gq3GEak+1yTXwd9c/8fkjEdz4Y0nitSr9Trph4r58FBEXFAF9mCRru+TAfaQZMAkE1PmWqXj0jF55zlgM8wlcegyQdSEkW5EHfhJWk7IBybQqEjdqpIX2KGcNJVf/gswzHlMAjonIVcklg ...
```
Missing copyright notice.

ppalucki replied:
This file is generated from script - I'll add license notice to to generation script

mmucek95 commented on hacks/delete-kind-local-shoot.sh:
```diff
@@ -0,0 +1,2 @@
+kubectl -n garden-local annotate shoot local "confirmation.gardener.cloud/deletion=true" --overwrite
```
Missing copyright notice.

ppalucki replied:
we won't be licensiny hacks/script with calling kubectl :) not IP there at all :)

mmucek95 commented on hacks/export-kubeconfig-for-local-shoot.sh:
```diff
@@ -0,0 +1,4 @@
+# source me
```
Missing copyright notice.

ppalucki replied:
done

mmucek95 commented on hacks/reconcile-trigger.sh:
```diff
@@ -0,0 +1 @@
+kubectl -n garden-local annotate shoot local "gardener.cloud/operation=reconcile" --overwrite
```
Missing copyright notice.

ppalucki replied:
too simple - I'm not adding License notice to signle commad kubectl scripts :)

mmucek95 commented on pkg/actuator/actuator.go:
```diff
@@ -0,0 +1,201 @@
+		return err
+	}
+
+	// Don't reconcile if already there ?! or we want to update with new configs/scripts?
```
To be answered before merging. If there is a new config/script it would make sense to reconcile.

ppalucki replied:
Yes I agree -  I moved all this considerations to issue when we can improve/optimise reconilation loop - the code is not best place to keeps such code snippets

Here you have: https://github.com/intel/gardener-extension-cri-resmgr/issues/43

but anyway done by removing

mmucek95 commented on pkg/actuator/actuator.go:
```diff
@@ -0,0 +1,201 @@
+		return err
+	}
+
+	// if err := a.deployDaemonsetToUninstallCriResMgr(ctx, ex); err != nil {
```
No longer needed

ppalucki replied:
removed !

mmucek95 commented on pkg/configs/configs-fixtures/bar:
```diff
@@ -0,0 +1 @@
+baz
```
A missing line at the end of the file.

ppalucki replied:
this is on purpose - because it as body of configuration file that I'm later checking in unittests

but fixed - I added a line + updated the unittests

mmucek95 commented on pkg/configs/configs.go:
```diff
@@ -0,0 +1,97 @@
+package configs
```
Missing copyright notice.

ppalucki replied:
added

mmucek95 commented on pkg/configs/configs.go:
```diff
@@ -0,0 +1,97 @@
+	// Has to be empty to allow helm values to merge
+	if providerConfig != nil {
+		if err := json.Unmarshal(providerConfig.Raw, &criResMgrConfig); err != nil {
+			// gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
```
Can we remove this unused code?

ppalucki replied:
removed and properaly returned an error instead of panicking

mmucek95 commented on pkg/configs/configs_test.go:
```diff
@@ -0,0 +1,84 @@
+package configs_test
```
Missing copyright notice.

ppalucki replied:
done

mmucek95 commented on charts/cri-resmgr-installation/templates/agent-adjustment-schema.yaml:
```diff
@@ -0,0 +1,85 @@
+apiVersion: apiextensions.k8s.io/v1
```
Missing copyright notice.

ppalucki replied:
added

ppalucki commented on Makefile:
```diff
@@ -11,27 +12,36 @@
+	go test -c -v ./pkg/actuator -o ./gardener-extension-cri-resmgr.actuator.test
+
+test:
+	# Those tests (renders charts, uses env to read files) change CWD during execution, required becuse realy on charts and fixtures
```
fixed

ppalucki commented on charts/cri-resmgr-installation/templates/installaton-configs.yaml:
```diff
@@ -0,0 +1,13 @@
+{{ range $config_name, $config_contents := .Values.configs }}
```
done

ppalucki commented on charts/gardener-extension-cri-resmgr/templates/configmap-imagevector-overwrite.yaml:
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.imageVectorOverwrite }}
```
added

ppalucki commented on charts/images.yaml:
```diff
@@ -0,0 +1,10 @@
+# Default values for image location internal installation and agent. Use localhost:5001 (local provider registry) until images will be published.
```
added

ppalucki commented on cmd/gardener-extension-cri-resmgr/app/app.go:
```diff
@@ -0,0 +1,116 @@
+				return fmt.Errorf("error completing options: %s", err)
+			}
+
+			// TODO: Flags version to allow override leader and
```
removed

ppalucki commented on cmd/gardener-extension-cri-resmgr/app/app.go:
```diff
@@ -0,0 +1,116 @@
+				MetricsBindAddress: "0",
+			}
+
+			// mgrOpts.ClientDisableCacheFor = []client.Object{
```
done

ppalucki commented:
@mmucek95 I addressed most of the comments in this commit - you may take a look: https://github.com/intel/gardener-extension-cri-resmgr/pull/33/commits/2cf9e056c353602b25831a3e1f9480161b258361

ppalucki commented on .vscode/launch.json:
```diff
@@ -0,0 +1,37 @@
+            "hideSystemGoroutines": true,
+        }
+    ]
+}
```
done

ppalucki commented:
marking tests as done - will be done on master when merged before release
moving "default config" work out of this PR https://github.com/intel/gardener-extension-cri-resmgr/issues/44

mmucek95 commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# CRI-Resource-Manager extension for Gardener*
+xw# CRI-Resource-Manager extension for Gardener*
```
This "xw" is supposed to be there?
